networkx/algorithms/similarity.py

```diff
@@ -1505,6 +1507,121 @@ def _simrank_similarity_numpy(
+    G,
+    source,
+    path_length=5,
+    c=0.5,
+    delta=0.1,
+    eps=None,
+    weight="weight",
+    remove_isolates=True,
+    k=None,
+):
+    """Common preparation code for Panther similarity algorithms.
+
+    Parameters
+    ----------
+    G : NetworkX graph
+        A NetworkX graph
+    source : node
+        Source node for similarity calculation
+    path_length : int
+        How long the randomly generated paths should be
+    c : float
+        A universal constant that controls the number of random paths to generate
+    delta : float
+        The probability parameter for similarity approximation
+    eps : float or None
+        The error bound for similarity approximation
+    weight : string or None
+        The name of an edge attribute that holds the numerical value used as a weight
+    remove_isolates : bool
+        Whether to remove isolated nodes from graph processing
+    k : int or None
+        The number of most similar nodes to return. If provided, validates that
+       ``k`` is not greater than the number of nodes in the graph.
+
+    Returns
+    -------
+    PantherPaths
+        A named tuple containing the prepared data:
+        - G: The graph (possibly with isolates removed)
+        - inv_node_map: Dictionary mapping node names to indices
+        - node_map: Array or list of node names
+        - sample_size: Number of paths to generate
+        - index_map: Populated index map of paths
+        - inv_sample_size: Inverse of sample size (for fast calculation)
+        - eps: Error bound for similarity approximation
+    """
+    import numpy as np
+
+    if source not in G:
+        raise nx.NodeNotFound(f"Source node {source} not in G")
+
+    isolates = set(nx.isolates(G))
+
+    if source in isolates:
+        raise nx.NetworkXUnfeasible(
+            f"Panther similarity is not defined for the isolated source node {source}."
+        )
+
+    if remove_isolates:
+        G = G.subgraph(node for node in G if node not in isolates).copy()
+
+    # According to [1], they empirically determined
+    # a good value for ``eps`` to be sqrt( 1 / |E| )
+    if eps is None:
+        eps = np.sqrt(1.0 / G.number_of_edges())
+
+    num_nodes = G.number_of_nodes()
+
+    # Check if k is provided and validate it against the number of nodes
+    if k is not None and not remove_isolates:  # For panther_vector_similarity
+        if num_nodes < k:
+            raise nx.NetworkXUnfeasible(
+                f"The number of requested nodes {k} is greater than the number of nodes {num_nodes}."
+            )
+
+    inv_node_map = {name: index for index, name in enumerate(G)}
+
+    # Calculate the sample size ``R`` for how many paths
+    # to randomly generate
+    t_choose_2 = math.comb(path_length, 2)
+    sample_size = int((c / eps**2) * (np.log2(t_choose_2) + 1 + np.log(1 / delta)))
+    index_map = {}
+
+    # Generate the random paths and populate the index_map
+    for _ in generate_random_paths(
```

Comment:
This call doesn't use a seed and makes the `panther*similarity` functions non-deterministic. Do we want them to be? Would it make sense to always explicitly use the same seed here, or to let the user specify a seed when calling `panther*similarity` (using an additional argument to the function)?
I'm asking because some of the tests occasionally fail if the random state is not set explicitly beforehand. I've indicated the ones I broke below, but there might be others.

If you think it's better to fix random seeds in our tests (as opposed to exposing them to users or setting them in the code), I think setting `np.random.seed(42)` works everywhere.

Reply:
Yes -- let's make sure that any place that uses random number generators has a seed in the function signature (with the associated random decorator on the function) and that it passes that `seed` parameter to all internally called functions.  

Thanks @Peiffap !!

Reply:
Think I got everything in my two most recent commits https://github.com/networkx/networkx/pull/4400/commits/4a1ee39cca75cbcfbc030341ceb44f0ccc4b6d29 and https://github.com/networkx/networkx/pull/4400/commits/79e43c68677f6db2a0d5a2cb2f5e5be194bad6da.